app/models/git_repository.rb

```diff
@@ -26,91 +27,135 @@ def refresh
   end
 
   def branch_info(name)
-    refresh unless repo
+    ensure_refreshed
     branch = git_branches.detect { |item| item.name == name }
     raise "Branch #{name} not found" unless branch
     branch.attributes.slice(*INFO_KEYS)
   end
 
   def tag_info(name)
-    refresh unless repo
+    ensure_refreshed
     tag = git_tags.detect { |item| item.name == name }
     raise "Tag #{name} not found" unless tag
     tag.attributes.slice(*INFO_KEYS)
   end
 
+  def entries(ref, path)
```

Comment:
Looking at the one places this is used, we will always be "directory scanning", and not looking at the DB for a cache of these entries (aka:  playbooks), correct?  Just making sure I understand correctly.

Reply:
This will emit the list of entries in a bare repo (no checkout)...that is, it uses Rugged methods that look at the git tree object.

Reply:
I think you missed the point of the question...

[libgit2](https://github.com/libgit2/libgit2) (which is what rugged uses under the hood), will be making file system calls to determine what playbooks are available to the user, and not "caching" the results in the database.

This means, if I am not on the `EmbeddedAnsible` appliance in a multi-appliance setup, I can't look for the available playbooks in a repo unless I repository checked out locally, correct?

Reply:
For what it is worth this can be addressed later, but just wanted to make sure that this is a known issue or I am missing a key concept here.

Reply:
I'm not sure I understand your distinction, but I thought "bare repo" made it clear that a) it's *not* scanning the filesystem for files but instead looking at the internal git graph (and yes, also not caching) and b) yes, it will need to ultimately checkout the files when they are run (i.e. the checkout method)

Reply:
@Fryguy I looked back at the files again and noticed that within `GitRepository.entries` (not here in `ConfigurationScriptSource`, there is a `ensure_refreshed` call, which would remove most of my concern.

* * *

But to clarify my statements around "directory scan", I am not trying to suggest that the `libgit2` is effectively doing a`Dir.glob(*.yml)` in the root directory, but it does have to read files in the `.git/` dir of the repository checkout location to determine what is available, so I am assuming there is some form FileIO happening.  I realize that using :v: directory scanning :v: may have been a poor choice of words, but I am 99% `libgit2` is not doing voodoo magic to determine current files for the `git` ref, and is just "reading files in `.git/`" to determine the "graph".

This means the `ensure_checkout` is required for `.entries` to function (thus, not a distributed solution), so if we to say... populate a dropdown in a UI only appliance that doesn't have the `EmbeddedAnsible` role enabled, it will have fetch the configured repo only to read the entries for that ref.  And if say the hosting platform for that repo is on, I don't know... github.com and it happens to be (\*cough\* \*cough\*) down when said drop down is being populated for the first time, then it could cause a confusing error in the UI (at the very least) since it can't determine without the (bare) repo present.

I am trying to point out an area of concern, not something that needs to be fixed right now.

Reply:
Hopefully GitHub this as "resolved", but wanted to followup with an answer to my concern (which was misguided):

After doing some digging and comparing to the `AnsibleTower` repo, I realized that this method is only called as part of the "refresh", and isn't actually called directly by other components (aka the UI).  The refresh in `ConfigurationScriptSource#sync` then sets payload records (which are the playbooks) and that is used by other workers to determine what playbooks are available.

So in short:  `.entries` is used just for the refresh, and should only be called on the `EmbeddedAnsible` role, and `ConfigurationScriptSource#configuration_script_payloads` is what will display items in the UI and elsewhere.  I had missed the code around that part of the `#sync`, and assumed `.entries` was being used everywhere.